modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+
+	buckets []float64
+
+	traceIDLabelName string
+}
+
+type nativeHistogramSeries struct {
+	// labels should not be modified after creation
+	labels        LabelPair
+	promHistogram prometheus.Histogram
+	lastUpdated   *atomic.Int64
+	histogram     *dto.Histogram
+}
+
+var (
+	_ Histogram = (*nativeHistogram)(nil)
+	_ metric    = (*nativeHistogram)(nil)
+)
+
+func newNativeHistogram(name string, buckets []float64, onAddSeries func(uint32) bool, onRemoveSeries func(count uint32), traceIDLabelName string) *nativeHistogram {
+	if onAddSeries == nil {
+		onAddSeries = func(uint32) bool {
+			return true
+		}
+	}
+	if onRemoveSeries == nil {
+		onRemoveSeries = func(uint32) {}
+	}
+
+	if traceIDLabelName == "" {
+		traceIDLabelName = "traceID"
+	}
+
+	return &nativeHistogram{
+		metricName:       name,
+		series:           make(map[uint64]*nativeHistogramSeries),
+		onAddSerie:       onAddSeries,
+		onRemoveSerie:    onRemoveSeries,
+		traceIDLabelName: traceIDLabelName,
+		buckets:          buckets,
+	}
+}
+
+func (h *nativeHistogram) ObserveWithExemplar(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) {
+	hash := labelValueCombo.getHash()
+
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	s, ok := h.series[hash]
+	if ok {
+		h.updateSeries(s, value, traceID, multiplier)
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
+		return
+	}
+
+	newSeries := h.newSeries(labelValueCombo, value, traceID, multiplier)
+	h.series[hash] = newSeries
+}
+
+func (h *nativeHistogram) newSeries(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) *nativeHistogramSeries {
+	newSeries := &nativeHistogramSeries{
+		// TODO move these labels in HistogramOpts.ConstLabels?
+		labels: labelValueCombo.getLabelPair(),
+		promHistogram: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    h.name(),
+			Help:    "Native histogram for metric " + h.name(),
+			Buckets: h.buckets,
+			// TODO check if these values are sensible and break them out
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 15 * time.Minute,
+		}),
+		lastUpdated: atomic.NewInt64(0),
+	}
+
+	h.updateSeries(newSeries, value, traceID, multiplier)
+
+	return newSeries
+}
+
+func (h *nativeHistogram) updateSeries(s *nativeHistogramSeries, value float64, traceID string, multiplier float64) {
+	for i := 0.0; i < multiplier; i++ {
+		s.promHistogram.(prometheus.ExemplarObserver).ObserveWithExemplar(
+			value,
+			map[string]string{h.traceIDLabelName: traceID},
+		)
+	}
+	s.lastUpdated.Store(time.Now().UnixMilli())
+}
+
+func (h *nativeHistogram) name() string {
+	return h.metricName
+}
+
+func (h *nativeHistogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	labelsCount := 0
+	if h.series[0] != nil {
```

Comment:
`h.series` is a map of hashes to histograms. won't this basically always return nil?

Reply:
Hmm, I'm not sure when this would be non-nil, but it doesn't seem that we test for this condition.

Reply:
Oof, yeah that's true. This is also a bug that exists in the original histogram code then 😬 The intention is "grab _any_ series in our map and check how many labels they have". But using `[0]` obviously isn't correct.

That this doesn't work isn't a huge deal, it just means the slice we allocate will be too small and gets reallocated as we add more labels (at most once per collection).

Reply:
I checked the hash function, and there is a 0 value returned, but it doesn't seem useful here to check since the label names would also be empty and zero length.  I'll drop this shortly.